flink-core/src/main/java/org/apache/flink/api/common/typeutils/TypeSerializerSnapshotSerializationUtil.java

```diff
@@ -145,8 +140,12 @@ public void read(DataInputView in) throws IOException {
                     serializerSnapshot = deserializeV2(in, userCodeClassLoader);
                     break;
                 case 1:
-                    serializerSnapshot = deserializeV1(in, userCodeClassLoader, serializer);
-                    break;
+                    throw new IOException(
+                            String.format(
+                                    "No longer supported version [%d] for TypeSerializerSnapshot. "
+                                            + "Please migrate away from the old TypeSerializerConfigSnapshot "
+                                            + "and use Flink 1.16 for the migration",
+                                    version));
```

Comment:
I find the way we check the incompatibility confusing atm.
There are at least two places that throw a similar exception:
1. `VersionedIOReadeableWritable` from `getIncompatibleVersionError`
2. Here from the `read` method.

I think 1) is actually a dead code now, as we tell in `TypeSerializerSnapshotSerializationProxy` that we support version 1.

Moreover I find the method `KeyedBackendSerializationProxy#getIncompatibleVersionError` misleading. I think at least the `Ops, this should not happen...` is unnecessary. You already get a message from `VersionedIOReadableWritable` that is an unsupported version. This comment does not give us any additional context. 

Lastly if you leave the `getIncompatibleVersionError` method I'd rename it something along the lines of "extra information", e.g. `getAdditionalDetailsForIncompatibleVersion`.

Reply:
Could we also clean this class a bit? I believe `org.apache.flink.api.common.typeutils.TypeSerializerSnapshotSerializationUtil.TypeSerializerSnapshotSerializationProxy#serializer`:
1. is never accessed
2. always null (apart from tests)

Reply:
I got rid of the "Ops, this should not happen...".  Per offline discussion we decided to keep this error message, but to reimplement it using `getCompatibleVersions` instead.